test/test_util.go

```diff
@@ -0,0 +1,174 @@
+// It is called when we want to test ReadOnly path
+// Creates object with default AWS session that presumably has PutObject permission
+func CondCreateObject(t *testing.T, awsRegion string, bucket string, body string, obj ObjectTestCase) {
+	// When encryption isn't set, trying to upload with default AWS session would fail as well
+	// hence the check here
+	if obj.expectPassWrite || obj.encryption == "" {
+		return
+	}
+
+	sess, err := aws.NewAuthenticatedSession(awsRegion)
+	if err != nil {
+		assert.FailNow(t, "Failed in creating session")
+	}
+	uploader := s3manager.NewUploader(sess)
+
+	logger.Log(t, fmt.Sprintf("Default Credential: uploading object %s...", obj.key))
+
+	_, err = UploadObjectWithUploaderE(
+		awsRegion,
+		bucket,
+		obj.key,
+		obj.encryption,
+		body,
+		uploader,
+	)
+
+	require.NoError(t, err, "Error raised when trying to upload objects that would be tested in ReadOnly paths")
+
+	logger.Log(t, fmt.Sprintf("Uploaded object %s with default credential", obj.key))
+
+}
+
+func UploadObjectWithUploaderE(awsRegion string, bucketName string, key string, encryption string, body string, uploader *s3manager.Uploader) (*s3manager.UploadOutput, error) {
+	s3Input := &s3manager.UploadInput{
+		Bucket: &bucketName,
+		Key:    &key,
+		Body:   strings.NewReader(body),
+	}
+
+	if encryption != "" {
+		s3Input.ServerSideEncryption = &encryption
+	}
+
+	return uploader.Upload(s3Input)
+
+}
+
+func GetS3ObjectWithSessionE(t *testing.T, bucket string, key string, sess *session.Session) (string, error) {
+	s3Client := s3.New(sess)
+
+	res, err := s3Client.GetObject(&s3.GetObjectInput{
+		Bucket: &bucket,
+		Key:    &key,
+	})
+
+	if err != nil {
+		return "", err
+	}
+
+	buf := new(bytes.Buffer)
+	_, err = buf.ReadFrom(res.Body)
+	if err != nil {
+		return "", err
+	}
+
+	contents := buf.String()
+	logger.Log(t, fmt.Sprintf("Read contents from s3://%s/%s", bucket, key))
+
+	return contents, nil
+}
+
+func validatePutObject(t *testing.T, awsRegion string, bucket string, obj ObjectTestCase, body string, sess *session.Session) {
+	_, err := retry.DoWithRetryInterfaceE(t,
+		"Trying to upload S3 Object..",
+		4, 3*time.Second,
+		func() (interface{}, error) {
+			return UploadObjectWithUploaderE(awsRegion, bucket, obj.key, obj.encryption, body, s3manager.NewUploader(sess))
+		})
+
+	if obj.expectPassWrite {
+		logger.Log(t, fmt.Sprintf("Try to upload using module permissions: %s", obj.key))
+		require.NoError(t, err)
+		logger.Log(t, fmt.Sprintf("%s uploaded using module permissions", obj.key))
+	} else {
+		require.Error(t, err)
+		logger.Log(t, fmt.Sprintf("%s could not be uploaded using module permissions", obj.key))
+	}
+}
+
+func validateGetObject(t *testing.T, bucket string, obj ObjectTestCase, expectedBody string, sess *session.Session) {
+	body, err := retry.DoWithRetryE(t,
+		"Trying to read S3 Object. We may not have permission or just be waiting it to be uploaded and available",
```

Comment:
Isn't s3 strongly consistent? Is there still a delay in availability after a put?
https://aws.amazon.com/blogs/aws/amazon-s3-update-strong-read-after-write-consistency/

Reply:
Regarding strongly consistency you are right! My comment was a bit outdated there... In addition to that, the `PutObject` test routine which comes first in the code will always wait (for IAM updates).
Therefore this is unnecessary now. Nice catch!

Reply:
fixed in https://github.com/hbarros-caylent/terraform-aws-s3/commit/8c3139a3c5797d3c40657f580669449564281e06